Config v2.0.0

OSadovy opened the pull request:
## Issue

https://github.com/harmony-one/bounties/issues/34

## Test

### Unit Test Coverage

Before:

```
ok      github.com/harmony-one/harmony/cmd/harmony      0.285s  coverage: 46.0% of statements
```

After:

```
ok      github.com/harmony-one/harmony/cmd/harmony      0.283s  coverage: 47.1% of statements
ok      github.com/harmony-one/harmony/cmd/harmony/config       0.267s  coverage: 36.5% of statements
```

### Test/Run Logs

```
root@07475f39e48e:~/go/src/github.com/harmony-one/harmony# bin/harmony -c testconfig.toml
Loaded config version 1.0.4 which is not latest (2.0.0).
Would you like to upgrade your config to the latest version? (y/n): y
previous config was stored at testconfig.toml.old
ERROR when loading bls key: stat ./.hmy/blskeys: no such file or directory
```

## Operational Checklist

1. **Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol?**. (If no, skip to question 8.)

    **NO**

8. **Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol?** (If no, continue to question 11.)

    **NO**

11. **Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage?**
    **NO**

JackyWYX commented on cmd/harmony/config/config.go:
```diff
@@ -255,67 +255,102 @@ func parseNetworkType(nt string) nodeconfig.NetworkType {
+	var config HarmonyConfig
+	// lexicographicaly compare
+	if header.Version <= v1TOMLConfigVersion {
+		config, err = loadHarmonyConfigV1(b)
```
It would be better to `loadHarmonyConfigV1` returns `harmonyConfigV1` and then parse V1 to V2 explicitly here. Which will increase the readability as well as give convenience for future updates.

OSadovy replied:
My idea is that we should have a file per old config version that we want to support, e.g. v1.go, v2.go etc, which should contain the old struct together with the code to convert that struct into the latest HarmonyConfig. that way we do not clutter main load function with every possible converting logic, but rather have that logic contained close to its struct.
I am not sure that having this convert logic for each version in the one place will increase readability, considering that we might have a number of versions in the future.
Or maybe I am missing the point, then correct me, please.

JackyWYX replied:
I am very glad to say that I definitely agree every point you have in the comment. What I have in mind is that, currently `LoadHarmonyConfig` is actually doing two jobs:

1. Load the config from toml file to its own version.
2. Parse the config to the latest version.

So I would love to see this two jobs are defined in two functions:

1. In each version file, load `toml` file to its own version. This will also handle minor version updates, E.g. v1.0.4 to v1.0.5.
2. Create a file `parse.go` to parse the legacy version to the latest.
```
// v1.go
func loadHarmonyConfigV1(file string) harmonyConfigV1 {}

// v2.go
func loadHarmonyConfigV2(file string) harmonyConfigV2 {}

// parse.go
func parseConfigToLatest(version vesion.Version, config interface{}) HarmonyConfig {
    switch version {
    ...
    }
}

func parseConfigV1ToV2(v1 harmonyConfigV1) harmonyConfigV2 {...}
func parseConfigV2ToV3(v2 harmonyConfigV2) harmonyConfigV3 {...}
```

In this case, when a new version of harmonyConfig is launched, there will be minimum update in the legacy file and functions.

JackyWYX commented on cmd/harmony/config/config.go:
```diff
@@ -255,67 +255,102 @@ func parseNetworkType(nt string) nodeconfig.NetworkType {
+	if header.Version <= v1TOMLConfigVersion {
+		config, err = loadHarmonyConfigV1(b)
+	} else {
+		err = toml.Unmarshal(b, &config)
```
Refactor this to `loadHarmonyConfigV2` would increase readability IMO.

OSadovy replied:
without renaming the struct itself I don't see a point in renaming/refactoring this one into standalone func. the whole idea is to have the latest and greatest HarmonyConfig and related functions public and exported, and other legacy structs and code to work with them private and hidden from other code. 

this way the action list for adding a new (backward incompatible) config version would be:
* copy current struct into private struct with version name in it and put it into a standalone file, say v2.go
* change HarmonyConfig to represent the newest version of config
* add a private loadV2Config() function to v2.go to convert old to new, and call it in the LoadHarmonyConfig publicly exported function.

if you have another plan in mind, could you please elaborate how do you see an action list for changing config to the new incompatible version would be with it?

JackyWYX replied:
> * copy current struct into private struct with version name in it and put it into a standalone file, say v2.go
> * change HarmonyConfig to represent the newest version of config
> * add a private loadV2Config() function to v2.go to convert old to new, and call it in the LoadHarmonyConfig publicly exported function.

Yes, you nailed it! 

And here I say refactor is just to refactor this one line `toml.Unmarshal` to a function to increase readability

JackyWYX commented on cmd/harmony/config/config.go:
```diff
@@ -255,67 +255,102 @@ func parseNetworkType(nt string) nodeconfig.NetworkType {
+func BackupAndUpgradeConfigToTheLatestVersion(cfg HarmonyConfig, fileName string) (string, error) {
+	oldFileName := fileName + ".old"
+	for {
+		_, err := os.Stat(oldFileName)
```
Potential forever loop and eats up all CPU resources. Would you please add a timely check logic and timeout here?

JackyWYX replied:
Do you have any thoughts on this one?

JackyWYX commented on cmd/harmony/config/config_test.go:
```diff
@@ -0,0 +1,302 @@
+}
+
+func TestV1_0_0Config(t *testing.T) {
+	testConfig := `Version = "1.0.4"
```
Would you please make this data a file and add it to test data folder? This will help with future updates.

JackyWYX commented on cmd/harmony/config/default.go:
```diff
@@ -0,0 +1,213 @@
+const TOMLConfigVersion = "2.0.0"
+
+const (
+	DefNetworkType = nodeconfig.Mainnet
```
Make items in this file private?

OSadovy replied:
They are used in flags_test.go unfortunately. another option that I considered is to move flags-related code into config package as well, but I declined the idea, since flags are not 100% following the config.

JackyWYX replied:
Ah I see. Please just keep it as it is, thanks

JackyWYX commented on cmd/harmony/main.go:
```diff
@@ -157,37 +194,85 @@ func raiseFdLimits() error {
+	if !terminal.IsTerminal(syscall.Stdin) {
+		return false
+	}
+	for {
```
Would you please add a timeout mechanism if possible?

OSadovy replied:
do you want for this prompt to disappear in some time if user does not take any action? Frankly speaking, I never saw this kind of behavior in cli applications before. or do you want me to limit number of attempts that use can mistype their choice?

IMHO if we know for sure that user is running this in the terminal, this current behavior is what one might expect for interactive cli application. but I am opened to your suggestions, if you elaborate, please.

JackyWYX replied:
There is a legacy way of starting harmony node with shell script `node.sh`: https://github.com/harmony-one/harmony/blob/main/scripts/node.sh

The suggested way of running harmony at that time was to run the shell script within tmux session. Though starting with `node.sh` is not recommended since a long time ago, we still spot some users keeps running nodes with deprecated `node.sh`. Thus the prompt without a timeout will potentially cause trouble in this case. 

Though it's uncommon, I would still suggest you to add a timeout mechanism if possible. Thanks!

JackyWYX commented on cmd/harmony/main.go:
```diff
@@ -157,37 +194,85 @@ func raiseFdLimits() error {
 
-func getHarmonyConfig(cmd *cobra.Command) (harmonyConfig, error) {
+func promptForConfigUpgrade() bool {
+	if !terminal.IsTerminal(syscall.Stdin) {
```
I think it will be more readable if we can put this if statement in upper function calls. What do you think?

JackyWYX commented on cmd/harmony/main.go:
```diff
@@ -157,37 +194,85 @@ func raiseFdLimits() error {
+		return config.HarmonyConfig{}, err
 	}
-	if config.Version != defaultConfig.Version {
+	if cfg.Version != config.DefaultConfig.Version {
```
Refactor to `config.IsLatestVersion(versionStr string)`? Also shall we consider the config version larger than current version?

OSadovy replied:
great idea! What if I add IsLatestVersion() method on the HarmonyConfig struct itself?

JackyWYX reviewed (commented):
Really appreciate for taking up this bounty. Added quite a few comments. Other than that, is there going to be flag changes as well? Also, do you mind adding some more detailed test scenario and reports please?

Much appreciated.

OSadovy commented on cmd/harmony/config/bls.go:
```diff
@@ -15,8 +15,8 @@ var (
 
-// setupConsensusKeys load bls keys and set the keys to nodeConfig. Return the loaded public keys.
-func setupConsensusKeys(hc harmonyConfig, config *nodeconfig.ConfigType) multibls.PublicKeys {
+// SetupConsensusKeys load bls keys and set the keys to nodeConfig. Return the loaded public keys.
```
I thought this functionality is config related (it is converting config-related data structure) and so belongs here. which place do you think is more appropriate? is cmd/harmony/ (the old place) better?

JackyWYX replied:
As far as I see, this file is more than parsing BLS config data structure. I would suggest you to put it back in `cmd/harmony`

OSadovy commented on cmd/harmony/config/config.go:
```diff
@@ -7,53 +7,48 @@ import (
 )
 
-// harmonyConfig contains all the configs user can set for running harmony binary. Served as the bridge
+// HarmonyConfig contains all the configs user can set for running harmony binary. Served as the bridge
```
actually, I think that it is better to have the latest config always to be named the same (e.g. HarmonyConfig, not HarmonyConfigV2). this will keep the fact that there might be several config versions private to the config package, and also keep the possible code changes for migrating to new config versions (v3, v4 etc) to the single package.

JackyWYX replied:
Yea, I totally agree with you about abstracting all details about config loading. What I have in mind is:

1. Latest config should always be named as `HarmonyConfig`
2. V2 should be named privately like `harmonyConfigV2`

Something like:

```
// config.go
type HarmonyConfig harmonyConfigV2

// v2.go
type harmonyConfigV2 struct {...}
```

Thus the two points you give in the comment will be still well kept:

1. Different versions are named privately.
2. Latest version is always called as `HarmonyConfig`

Moreover, his code will more explicitly show that the latest HarmonyConfig is of version V2, and will have smaller code changes when updating to a new version (Only need to redefine `type HarmonyConfig harmonyConfigV3` instead of rename the file to V2). 

Or do you have any other concerns about this? We can have a more in depth discussion.

OSadovy commented on cmd/harmony/config/config.go:
```diff
@@ -167,43 +160,55 @@ type syncConfig struct {
 
 	netType := config.Network.NetworkType
-	parsed := parseNetworkType(netType)
+	parsed := ParseNetworkType(netType)
```
yes, it is atleast used in flags.go in getNetworkType()

JackyWYX replied:
Ah I see. Nice!

OSadovy commented:
> Really appreciate for taking up this bounty. Added quite a few comments. 
Many thanks for your thorough review!

>Other than that, is there going to be flag changes as well? 
I changed some flags (sync.legacy.client -> dnssync.legacy.client, sync.legacy.server -> dnssync.legacy.server) with preserving old versions as well. what else flag changes would you like to see? i thought about changing dns.zone and dns.port, but because those flags are not 100% following config naming convention, I dropped the idea (not to change just for sake of changing).

>Also, do you mind adding some more detailed test scenario and reports please?
Do you mean describing what tests I performed manually, or adding more automatic test cases?

> Much appreciated.
Thanks!

JackyWYX commented on cmd/harmony/config/config.go:
```diff
@@ -255,67 +255,102 @@ func parseNetworkType(nt string) nodeconfig.NetworkType {
+	}
+	var config HarmonyConfig
+	// lexicographicaly compare
+	if header.Version <= v1TOMLConfigVersion {
```
I think hashicorp solution should be better IMO: https://github.com/hashicorp/go-version.

I understand that lexicographic solution will work for most of the cases if the version string is legit. But what about the case that user accidentally modified version string and don't notice the change? E.g. ` v2.0.0`. If this is the case, the config will be silently loaded as v1 and then parsed to v2, which might cause loaded config to have an unexpected result. 

So this is my concern about "hacking" with the string comparison. I would encourage a systematic approach to add the go-version package could potentially provide more stability to the system.

JackyWYX commented:
Thanks!

> I changed some flags (sync.legacy.client -> dnssync.legacy.client, sync.legacy.server -> dnssync.legacy.server) with preserving old versions as well. what else flag changes would you like to see? i thought about changing dns.zone and dns.port, but because those flags are not 100% following config naming convention, I dropped the idea (not to change just for sake of changing).

The legacy flags still need to be supported IMO, since some user may start harmony with some flags in systemd, and the harmony will not work for these legacy flags. I would suggest to copy the legacy flags with exactly the same names and make them backward compatible.
